### PR TITLE
Attempt at implementing gradient accummulation during training (with JAX backend)

### PR DESCRIPTION
**Pull request by ixig**
Keras3 currently does not support gradient accumulation for smoother gradient updates when memory constraints limit batch sizes. I believe being able to do this is very beneficial to many Keras users, as larger models are crimping batch sizes. This is very easy for users coming from PyTorch to do, as loss.backwards() accumulates gradients by default, and the PyTorch Lightning [trainer](https://lightning.ai/docs/pytorch/stable/common/trainer.html#accumulate-grad-batches) supports this training flag as well.

I previously submitted Issue #18923 and was asked to submit a PR with my code implementation. This PR is in response to that. This implementation is only for the JAX Trainer as i. it is only a POC ii. my interest is with the JAX backend as its speed is the most-compelling reason for switching from PyTorch.

I've tested the code changes in this PR to be functionally working as an implementation of gradient accumulation, but it's not ideal. I'm submitting it as a reference implementation to support #18923. The following are its known issues:

1. Creation of the gradient accumulation buffer variables is currently done in the `BaseOptimizer` class, because that class provides the `add_variable_from_reference()` method. The buffer should instead be created in `JAXTrainer` (but I didn't know how), then no changes to `BaseOptimizer` is necessary (nor is the awkward need to set `use_grad_accum=True` when instantiating an optimizer).

2. Conserving GPU memory is very important (it is the motivation behind this feature) -- I'm not sure why `donate_argnames="state"` in the jax.jit decorator results in a "Attempt to donate the same buffer twice" error.

3. I'm not familiar with the Keras internals, so my handling (creation/tracking/sharding/synchronization/etc.) of the `accum_grads` variables may not be the most performant.

4. As was suggested by @fchollet , my initial attempt was to implement this feature in `BaseOptimizer`, but I ran into the "Attempted boolean conversion of traced array" JAX compilation error when I tried to test the modulo value of `self.iterations` (which contains the batch iteration number) in `apply()`.

5. I got around this by implementing a `train_step_grad_accum()` in `JAXTrainer` that took a static argument `optimize` (selects between two execution paths). However, this results is a longer graphing/compilation time as both paths need to be traced.

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for opening the PR, it really helps ground the discussion.

> Conserving GPU memory is very important (it is the motivation behind this feature)

Can you say more about this? How is the feature expected to lead to memory savings?

> As was suggested by @fchollet , my initial attempt was to implement this feature in BaseOptimizer, but I ran into the "Attempted boolean conversion of traced array" JAX compilation error when I tried to test the modulo value of self.iterations (which contains the batch iteration number) in apply().

You cannot do `if x` where `x` in a JAX/TF tensor. You would have to use a `ops.cond` operation. Like:
```python
def my_apply(...):
   return ops.cond(
        ops.equal(ops.mod(self.iterations, self.accumulation_steps), 0),
        real_apply_fn,
        accumulation_fn,
   )
```

Would it make sense to have the feature implemented purely as an optimizer, where the optimizer would record past gradients in `apply()` and do nothing, but would actually update the variables using the accumulated gradients every N steps? It seems like it would be much simpler than modifying the Trainer class. It could also be implemented in a way that is backend-agnostic, avoid code duplication in each trainer.

If the `cond` issue was the only problem, then it should be doable?

**Comment by ixig** [comment]
> Can you say more about this? How is the feature expected to lead to memory savings?

Gradient accumulation (GA) does not in itself lead to memory savings, since memory has to be allocated for the accumulation buffer. What I meant is that the reason for a user wanting to use GA is to maximize the use of GPU memory for increasing model size/complexity -- at the expense of memory that's leftover for holding the batch data input, and scratchpad for the intermediate results of the forward/backward passes. In our models, we're often left with only enough memory for a batch size of 8 or 16, for example, which results in the inability to achieve optimal convergence due to noisy gradients (premature early-stopping, more epochs to converge, etc.). GA therefore is a mechanism to potentially allow for _net_ [larger batch sizes](https://arxiv.org/abs/1711.00489) than would otherwise be possible for available memory.

Results [depend](https://towardsdatascience.com/how-to-break-gpu-memory-boundaries-even-with-large-batch-sizes-7a9c27a400ce) on the model, but if the model parameter count is relatively small compared to the memory requirements for the batch and scratchpad data, then spending a bit of memory (equal to that consumed for storing the model parameters) for GA, particularly when `grad_accum_steps` is large (e.g. 4), pays off in allowing for net larger training batch sizes.

So my point is that GPU memory is precious and a good GA implementation must not be wasteful in its implementation to be worthwhile.

**Comment by fchollet** [comment]
I'll give the optimizer implementation a shot.

**Comment by fchollet** [comment]
There's actually one major obstacle in JAX (other backends are fine): JAX `cond` cannot have side effects. This means we cannot update variables in a conditional branch.

**Comment by AakashKumarNain** [comment]
> I'll give the optimizer implementation a shot.
Agreed. Implementing it at an optimzer level makes it backend agnostic and avoids the deduplication in the trainer. In fact, if we implement it at optimizer level, then it becomes easy for using it in custom loops as well

**Comment by innat** [comment]
> Would it make sense to have the feature implemented purely as an optimizer 

+1

https://github.com/andreped/GradientAccumulator
cc @andreped

**Comment by fchollet** [comment]
Also, there's additional complexity if you want to support tf.distribute in the TF backend. So overall if you want to be feature-complete this is very much non-trivial, however simple the initial implementation may look.

**Comment by fchollet** [comment]
Here's my PR: https://github.com/keras-team/keras/pull/18951/files

I ended up with a sizeable refactor of optimizers in general.

**Comment by andreped** [comment]
Note that AFAIK BatchNormalization needs to be handled with care, due to how it is updated in both the forward and backward steps. In GradientAccumulator, I developed a drop-in replacement:
https://github.com/andreped/GradientAccumulator/blob/main/gradient_accumulator/layers.py#L194

The idea is basically to accumulate the mean and standard deviations in the forward step with a `accum_steps` variable. This can be implemented by adding GA support to the BatchNormalization layer directly. I implemented a POC which seems to work as it should.
